File: p230m/kth_smallest_bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
const Tree = require('binary_tree');
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function (root, k) {
  let count = 0;
  const recInOrder = function (node) {
    if (node == null) return null;
    let lret = recInOrder(node.left);
    if (lret) return lret;
    if (++count === k) return node.val;
    return recInOrder(node.right);
  };
  return recInOrder(root);
};
// TEST
[
  ['1,#,#', 1, 1],
  ['2,1,#,#,3,#,#', 2, 2],
  ['2,1,#,#,3,#,#', 3, 3],
  ['5,3,2,#,#,4,#,#,6,#,7,#,#', 1, 2],
  ['5,3,2,#,#,4,#,#,6,#,7,#,#', 2, 3],
  ['5,3,2,#,#,4,#,#,6,#,7,#,#', 3, 4],
  ['5,3,2,#,#,4,#,#,6,#,7,#,#', 4, 5],
  ['5,3,2,#,#,4,#,#,6,#,7,#,#', 5, 6],
  ['5,3,2,#,#,4,#,#,6,#,7,#,#', 6, 7],
].forEach(([array, k, expected]) => {
  const tree = Tree.deserialize(array);
  const actual = kthSmallest(tree, k);
  console.log('The', k + '-th smallest value in BST', array, '->', actual);
  console.assert(actual === expected);
});

File: p230m/kth_smallest_bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from local_packages.binary_tree import TreeNode
from typing import Optional


class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        def inorder(node: TreeNode):
            if node:
                yield from inorder(node.left)
                yield node
                yield from inorder(node.right)

        for i, node in enumerate(inorder(root)):
            if i == k - 1:
                return node.val


# TESTS
for array, k, expected in [
    ["1,#,#", 1, 1],
    ["2,1,#,#,3,#,#", 2, 2],
    ["2,1,#,#,3,#,#", 3, 3],
    ["5,3,2,#,#,4,#,#,6,#,7,#,#", 1, 2],
    ["5,3,2,#,#,4,#,#,6,#,7,#,#", 2, 3],
    ["5,3,2,#,#,4,#,#,6,#,7,#,#", 3, 4],
    ["5,3,2,#,#,4,#,#,6,#,7,#,#", 4, 5],
    ["5,3,2,#,#,4,#,#,6,#,7,#,#", 5, 6],
    ["5,3,2,#,#,4,#,#,6,#,7,#,#", 6, 7],
]:
    sol = Solution()
    tree = TreeNode.deserialize(array)
    actual = sol.kthSmallest(tree, k)
    print("The", f"{k}-th smallest value in BST", array, "->", actual)
    assert actual == expected
